op-service/sources/l1_beacon_client.go

```diff
@@ -0,0 +1,173 @@
+		if timestamp < genesisTime {
+			return 0, fmt.Errorf("provided timestamp (%v) precedes genesis time (%v)", timestamp, genesisTime)
+		}
+		return (timestamp - genesisTime) / secondsPerSlot, nil
+	}
+	return cl.timeToSlotFn, nil
+}
+
+// GetBlobSidecars fetches blob sidecars that were confirmed in the specified L1 block with the
+// given indexed hashes. Order of the returned sidecars is not guaranteed, and blob data is not
+// checked for validity.
+func (cl *L1BeaconClient) GetBlobSidecars(ctx context.Context, ref eth.L1BlockRef, hashes []eth.IndexedBlobHash) ([]*eth.BlobSidecar, error) {
+	if len(hashes) == 0 {
+		return []*eth.BlobSidecar{}, nil
+	}
+	slotFn, err := cl.GetTimeToSlotFn(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to get time to slot function", err)
+	}
+	slot, err := slotFn(ref.Time)
+	if err != nil {
+		return nil, fmt.Errorf("%w: error in converting ref.Time to slot", err)
+	}
+
+	builder := strings.Builder{}
+	builder.WriteString(sidecarsMethodPrefix)
+	builder.WriteString(strconv.FormatUint(slot, 10))
+	builder.WriteRune('?')
+	v := url.Values{}
+	for i := range hashes {
+		v.Add("indices", strconv.FormatUint(hashes[i].Index, 10))
+	}
+	builder.WriteString(v.Encode())
+
+	var resp eth.APIGetBlobSidecarsResponse
+	if err := cl.apiReq(ctx, &resp, builder.String()); err != nil {
+		return nil, fmt.Errorf("%w: failed to fetch blob sidecars for slot %v block %v", err, slot, ref)
+	}
+	if len(hashes) != len(resp.Data) {
+		return nil, fmt.Errorf("expected %v sidecars but got %v", len(hashes), len(resp.Data))
+	}
+
+	return resp.Data, nil
+}
+
+// GetBlobs fetches blobs that were confirmed in the specified L1 block with the given indexed
+// hashes. The order of the returned blobs will match the order of `hashes`.  Confirms each
+// blob's validity by checking its proof against the commitment, and confirming the commitment
+// hashes to the expected value. Returns error if any blob is found invalid.
+func (cl *L1BeaconClient) GetBlobs(ctx context.Context, ref eth.L1BlockRef, hashes []eth.IndexedBlobHash) ([]*eth.Blob, error) {
+	blobSidecars, err := cl.GetBlobSidecars(ctx, ref, hashes)
+	if err != nil {
+		return nil, fmt.Errorf("%w: failed to get blob sidecars for L1BlockRef %s", err, ref)
+	}
+	return blobsFromSidecars(blobSidecars, hashes)
+}
+
+func blobsFromSidecars(blobSidecars []*eth.BlobSidecar, hashes []eth.IndexedBlobHash) ([]*eth.Blob, error) {
+	out := make([]*eth.Blob, len(hashes))
+	for i, ih := range hashes {
+		// The beacon node api makes no guarantees on order of the returned blob sidecars, so
+		// search for the sidecar that matches the current indexed hash to ensure blobs are
+		// returned in the same order.
+		var sidecar *eth.BlobSidecar
+		for _, sc := range blobSidecars {
+			if uint64(sc.Index) == ih.Index {
+				sidecar = sc
+				break
+			}
+		}
+		if sidecar == nil {
+			return nil, fmt.Errorf("no blob in response matches desired index: %v", ih.Index)
+		}
```

Comment:
```suggestion
		scIndex := slices.IndexFunc(blobSidecars, func(sc *eth.BlobSidecar) {return uint64(sc.Index) == ih.Index})
		if scIndex == -1 {
			return nil, fmt.Errorf("no blob in response matches desired index: %v", ih.Index)
		}
		sidecar := blobSidecars[scIndex]
```

Rough proposal to use the slices package. I'm not sure if my code is 100% correct, but I'm interested on your take on it vs manually iterating the array. Ideally there would be a `Find` and `FindFunc` functions in the slices package but those don't seem to exist.

Reply:
nice, had a few syntax issues but fixed & included.